### PR TITLE
Save platforms and plugins to devDependencies

### LINKED ISSUE #64: Save platforms and plugins to devDependencies
# Feature Request

## Motivation Behind Feature

Currently, Cordova platforms and plugins are saved as `dependencies` in the `package.json` file. Hence, during the Cordova prepare step, necessary files for plugins are moved for bundling with the app. These `dependencies` are not used directly with the app, so I believe, they should be treated as `devDependencies`.

I want to clear out Cordova platforms and plugins from `dependencies` because it could affect platforms that leverage the `npm` modules. This issues can be seen with the newest platform - Electron. With Electron it is expected that you use `npm` packages within the application and should be installed as `dependencies`.

For instance, if we want to use `lodash` with Electron app, we would `npm install lodash` and then import it using `require`. When we `run` the application everything works as expected because it uses the module from `{project}/node_modules`. But with `build` we need to take all the `dependencies` from the project and install them within the app. 

As a result:
 - With Cordova platforms and plugins defined in the `dependencies`, for example, `cordova-electron `the packaged app size will increase greatly.

- With Cordova platforms and plugins defined in the `devDependencies`, they would not be installed with the app and reduce the package size.

_Again, Cordova platforms and plugins should be treated as development building blocks and shouldn't be in the app_

## Feature Description

The current implementation of the npm arguments handling is the following:
```js
function npmArgs (target, userOptions) {
    const opts = Object.assign({ production: true }, userOptions);

    const args = ['install', target];

    if (opts.production) {
        args.push('--production');
    }
    if (opts.save_exact) {
        args.push('--save-exact');
    } else if (opts.save) {
        args.push('--save');
    } else {
        args.push('--no-save');
    }
    return args;
}
```

My suggestion would be instead of using `--save` flag, change it to `--save-dev`. While the `--save` option still appears to work, it is no longer required. The `npm install` saves any specified packages into `dependencies` by default.

Besides that, we also append `--production` flag. With the `--production` flag npm will not install modules listed in `devDependencies`. However, the `--production` flag has no particular meaning when we are adding a dependency to a project. I think that should be removed. Plus, if you use `--production` and `--save-dev` flags together, it causes `npm install` to break.

Later we could also remove options, like production, that get passed from `cordova-lib` to the `cordova-fetch` and do a general clean up.

## Alternatives or Workarounds

Manually move Cordova platforms and plugins, that don't need to be bundled with an app, from `dependencies` to the `devDependecies`.

### PR DESCRIPTION
**Pull request by GedasGa**
### Motivation and Context

 Save platforms and plugins to devDependencies and close #64.

### Description

For more details please refer to the #64 issue.

### Testing

`npm t`

### Checklist

- [x] I've run the tests to see all new and existing tests pass
- [x] I added automated test coverage as appropriate for this change
- [ ] Commit is prefixed with `(platform)` if this change only applies to one platform (e.g. `(android)`)
- [x] If this Pull Request resolves an issue, I linked to the issue in the text above (and used the correct [keyword to close issues using keywords](https://help.github.com/articles/closing-issues-using-keywords/))
- [ ] I've updated the documentation if necessary

### CODE REVIEW

**Comment by erisu** [comment]
@GedasGa, can you rebase and fix the conflicts?

File: index.js [hunk 1]
```diff
@@ -171,9 +167,9 @@ module.exports.uninstall = function (target, dest, opts) {
             // set the directory where npm uninstall will be run
             opts.cwd = dest;
 
-            // if user added --save flag, pass it to npm uninstall command
+            // if user added --save flag, pass --save-dev flag to npm uninstall command
             if (opts.save) {
-                fetchArgs.push('--save');
+                fetchArgs.push('--save-dev');
```
Comment (raphinesse) [inline index.js]: I think we shouldn't change this at all. `npm uninstall --save foo` will uninstall `foo` and remove it from either `dependencies` or `devDependencies`, wherever it was included.

Comment (GedasGa) [reply index.js]: Will it?

As per npm CLI documentation, it says:

> npm uninstall takes 3 exclusive, optional flags which save or update the package version in your main package.json:
> 
> -S, --save: Package will be removed from your dependencies.
> 
> -D, --save-dev: Package will be removed from your devDependencies.
> 
> -O, --save-optional: Package will be removed from your optionalDependencies.
> 
> --no-save: Package will not be removed from your package.json file.

Anyway, since we expect Cordova packages to be installed only as a devDependencies, I don't think there should be a problem.

Comment (erisu) [reply index.js]: I just ran a quick test myself and it does what @raphinesse says.

```
npm init
npm install -D cordova-common
npm uninstall --save cordova-common
```

The example above installs `cordova-common` as a `devDependencies`. `uninstall --save` removed `cordova-common` from `devDependencies` as well.

For a different test case:

```
npm init
npm install cordova-common
npm uninstall --save-dev cordova-common
```

The example above installs `cordova-common` not as a `devDependencies`. `uninstall --save-dev` removed `cordova-common` from `dependencies` as well even though the `--save-dev` flag was passed.

Basically, it seems the flag does make a difference.

Comment (breautek) [reply index.js]: I have never specified the kind of dependency when uninstalling packages. So yah, we could remove the flag from the uninstall command altogether

Comment (raphinesse) [reply index.js]: - Removing the flag might be unsafe as users could have specified 'no-save' in their `.npmrc`
- When I first read the code I found `uninstall --save-dev` to be irritating. I was unsure of it's behavior
- Then again, according to the docs cited by @GedasGa, we definitely _should_ have `--save-dev` and not `--save` here. Basically npm _could_ change the behavior observed by us in a patch release. I don't think they will, but still...
- But what we _actually want_ is the current behavior of `npm`: uninstalling and removing the package from `package.json` regardless of the dependency category. That way, old dependencies that might still be production ones are correctly removed too

So in conclusion, what does that mean? :sweat_smile:

Comment (raphinesse) [reply index.js]: I think the way you have it now is correct, @GedasGa :+1:
